lib/logger_splunk_backend.ex

```diff
@@ -5,140 +5,304 @@ defmodule Logger.Backend.Splunk do
 
   require Logger
 
-  @impl true
+  @impl :gen_event
   def init({__MODULE__, name}) do
     state = %{
       name: name,
       buffer: [],
       max_buffer: 32,
       buffer_size: 0,
-      connector: Logger.Backend.Splunk.Output.Http,
+      response: nil,
+      output: nil,
+      printing_response?: false,
+      error_device: :stdio,
+      erlang_host_json: ~s("host":""),
+      index_json: [],
       host: nil,
       level: :debug,
       format: @default_format,
       compiled_format: Logger.Formatter.compile(@default_format),
       metadata: [],
       token: ""
     }
+
     {:ok, configure([], state)}
   end
 
-  @impl true
-  def handle_call({:configure, opts}, state) do
-    {:ok, :ok, configure(opts, state)}
+  def init(__MODULE__) do
+    # use a default name
+    init({__MODULE__, __MODULE__})
   end
 
-  @impl true
-  def handle_call(:connector, %{connector: connector} = state) do
-    {:ok, {:ok, connector}, state}
+  @impl :gen_event
+  def handle_call({:configure, opts}, state) do
+    {:ok, :ok, configure(opts, state)}
   end
 
-  @impl true
-  def handle_event({level, _gl, {Logger, msg, ts, md}}, %{level: min_level} = state) do
-    state = if is_nil(min_level) or Logger.compare_levels(level, min_level) != :lt do
-      log_event(level, msg, ts, md, state)
-    else
+  @impl :gen_event
+  def handle_event({level, _gl, {Logger, msg, ts, md}}, state) do
+    %{level: min_level, buffer_size: buffer_size, max_buffer: max_buffer, response: response} =
       state
+
+    cond do
+      not meet_level?(level, min_level) ->
+        {:ok, state}
+
+      is_nil(response) ->
+        # we aren't in the middle of a request, so send the message right away
+        {:ok, log_event(level, msg, ts, md, state)}
+
+      buffer_size < max_buffer ->
+        # we are in the middle of a request, but there's still room to
+        # buffer. we'll send these messages when the current one completes,
+        # from handle_http_event/2.
+        {:ok, buffer_event(level, msg, ts, md, state)}
+
+      true ->
+        # we're in the middle of the request, but have reached our max buffer size
+        state = buffer_event(level, msg, ts, md, state)
```

Comment:
Should we still be buffering the event here if we've reached the max size? Looking at the definition of `buffer_event/5`, I don't see where it would reject the event, so I think we will continue to grow the buffer?

Reply:
You're correct, it does grow the buffer. The difference is that we wait for the buffer to be processed by `await_response`, instead of returning. The Logger keeps track of how many unprocessed messages there are in the message queue, so this will let them pile up in the background while we handle the HTTP response messages. If too many messages pile up in the queue, Logger will switch to sync mode and apply backpressure to clients, and eventually start discarding messages entirely.